File: app/routers/conversations.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from app.services import conversations as conversations_service

router = APIRouter(tags=["conversations"])


class CreateConversationPayload(BaseModel):
    title: str = Field(..., min_length=1)


class AddMessagePayload(BaseModel):
    role: str = Field(..., pattern="^(user|assistant)$")
    content: str = Field(..., min_length=1)
    sources: list | None = None
    agent_steps: list | None = None


class UpdateTitlePayload(BaseModel):
    title: str = Field(..., min_length=1)


@router.get("/conversations")
async def list_conversations():
    return {"conversations": conversations_service.list_conversations()}


@router.post("/conversations")
async def create_conversation(payload: CreateConversationPayload):
    return conversations_service.create_conversation(payload.title)


@router.get("/conversations/{conversation_id}")
async def get_conversation(conversation_id: str):
    conversation = conversations_service.get_conversation(conversation_id)
    if conversation is None:
        raise HTTPException(status_code=404, detail="Conversa não encontrada.")
    return conversation


@router.delete("/conversations/{conversation_id}")
async def delete_conversation(conversation_id: str):
    deleted = conversations_service.delete_conversation(conversation_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Conversa não encontrada.")
    return {"ok": True}


@router.post("/conversations/{conversation_id}/messages")
async def add_message(conversation_id: str, payload: AddMessagePayload):
    conversation = conversations_service.get_conversation(conversation_id)
    if conversation is None:
        raise HTTPException(status_code=404, detail="Conversa não encontrada.")

    return conversations_service.add_message(
        conversation_id=conversation_id,
        role=payload.role,
        content=payload.content,
        sources=payload.sources,
        agent_steps=payload.agent_steps,
    )


@router.put("/conversations/{conversation_id}")
async def update_conversation_title(conversation_id: str, payload: UpdateTitlePayload):
    conversation = conversations_service.update_conversation_title(
        conversation_id, payload.title
    )
    if conversation is None:
        raise HTTPException(status_code=404, detail="Conversa não encontrada.")
    return conversation

File: src/lib/api/conversations.ts
import { getApiBaseUrl } from './config';
import type {
	ConversationsListResponse,
	ConversationDetail,
	ConversationSummary,
	StoredMessage,
	FastApiError,
	AskSource,
	AskAgentStep
} from './types';

function parseErrorDetail(body: FastApiError): string {
	if (typeof body.detail === 'string') return body.detail;
	if (Array.isArray(body.detail) && body.detail.length > 0)
		return body.detail.map((item) => item.msg).join('; ');
	return 'Erro na operação.';
}

export async function listConversations(): Promise<ConversationSummary[]> {
	const response = await fetch(`${getApiBaseUrl()}/conversations`);
	if (!response.ok) throw new Error(`Erro ao listar conversas (${response.status}).`);
	const data = (await response.json()) as ConversationsListResponse;
	return data.conversations;
}

export async function getConversation(id: string): Promise<ConversationDetail> {
	const response = await fetch(`${getApiBaseUrl()}/conversations/${id}`);
	if (!response.ok) throw new Error(`Erro ao carregar conversa (${response.status}).`);
	return (await response.json()) as ConversationDetail;
}

export async function createConversation(title: string): Promise<ConversationSummary> {
	const response = await fetch(`${getApiBaseUrl()}/conversations`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ title })
	});
	if (!response.ok) throw new Error(`Erro ao criar conversa (${response.status}).`);
	return (await response.json()) as ConversationSummary;
}

export async function deleteConversation(id: string): Promise<void> {
	const response = await fetch(`${getApiBaseUrl()}/conversations/${id}`, { method: 'DELETE' });
	if (!response.ok) throw new Error(`Erro ao excluir conversa (${response.status}).`);
}

export async function addMessage(
	conversationId: string,
	role: 'user' | 'assistant',
	content: string,
	sources?: AskSource[],
	agentSteps?: AskAgentStep[]
): Promise<StoredMessage> {
	const response = await fetch(`${getApiBaseUrl()}/conversations/${conversationId}/messages`, {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ role, content, sources, agent_steps: agentSteps })
	});
	if (!response.ok) throw new Error(`Erro ao salvar mensagem (${response.status}).`);
	return (await response.json()) as StoredMessage;
}

export async function updateConversationTitle(id: string, title: string): Promise<void> {
	const response = await fetch(`${getApiBaseUrl()}/conversations/${id}`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ title })
	});
	if (!response.ok) throw new Error(`Erro ao atualizar conversa (${response.status}).`);
}
